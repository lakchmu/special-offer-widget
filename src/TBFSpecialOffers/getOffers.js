import request from './request';
import OffersStorage from './offersStorage';
import { API_END_POINT, API_METHOD_SPECIAL_OFFERS } from '../constants';

function getOffers(token, filters) {
  const url = API_END_POINT + API_METHOD_SPECIAL_OFFERS;
  return request(url, 'get', token)
    .then(function(response) {
      return response.json();
    })
    .then(json => {
      return new OffersStorage(json.special_offers);
    });
}

export default getOffers;