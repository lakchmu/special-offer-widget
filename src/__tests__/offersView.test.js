/* global jest, describe, it, expect */

import getOffers from '../TBFSpecialOffers/getOffers';
import OffersView from '../TBFSpecialOffers/offersView';

jest.mock('../TBFSpecialOffers/request');
jest.mock('../index.css');

function setup() {
  const data = {
    token: 'KO96pMbbIBTQPl7vnoJMXmIrEl2cYoUZS83QlL0Abu0',
    filters: {},
  };
  return getOffers(data.token, data.filters);
}

function getOptions(key) {
  let options;
  switch (key) {
    case 'empty':
      options = {};
      break;
    case 'normal-render-function':
      options = {
        renderTemplate: (offerModel) => {
          const { title } = offerModel;
          const template = `<div class="tbf-so-offer">\n<div class="tbf-so-offer__title">${title}</div>\n</div>`;
          return template;
        },
      };
      break;
    case 'not function':
      options = {
        renderTemplate: 'not function',
      };
      break;
    default:
      break;
  }
  return options;
}

function setupTest(rootElement, optionKey) {
  document.body.innerHTML = '<div class="special-offers"></div>';
  return setup()
    .then(offersStorage => new OffersView(rootElement, offersStorage, getOptions(optionKey)));
}

function setupWrapperTest(rootElement, optionKey) {
  document.body.innerHTML = '<div class="special-offers"></div>';
  return setup()
    .then(offersStorage => function testWrapper() {
      return new OffersView(rootElement, offersStorage, getOptions(optionKey));
    });
}

describe('Testing class methods', () => {
  it('Root element for widget does not exist', () => {
    expect.assertions(1);
    return setupWrapperTest('.not-existing-root-element', 'empty')
      .then(testWrapper => expect(testWrapper).toThrow());
  });

  it('Testing default render function', () => {
    expect.assertions(1);
    return setupTest('.special-offers', 'empty')
      .then(offersView => expect(offersView.rootElement).toMatchSnapshot());
  });

  it('Testing adding css class to root element', () => {
    expect.assertions(1);
    return setupTest('.special-offers', 'empty')
      .then((offersView) => {
        const containsClass = offersView.rootElement.classList.contains('tbf-so-special-offers');
        return expect(containsClass).toBeTruthy();
      });
  });
});

describe('Testing custom render function', () => {
  it('Testing rendering', () => {
    expect.assertions(1);
    return setupTest('.special-offers', 'normal-render-function')
      .then(offersView => expect(offersView.rootElement).toMatchSnapshot());
  });

  it('Custom render function is not a function', () => {
    expect.assertions(1);
    return setupWrapperTest('.special-offers', 'not function')
      .then(testWrapper => expect(testWrapper).toThrow());
  });
});

describe('Testing event listeners', () => {
  it('Testing click event by "More" button', () => {
    expect.assertions(4);
    return setupTest('.special-offers', 'empty')
      .then((offersView) => {
        const moreButtons = offersView.rootElement.querySelectorAll('.tbf-so-offer__more-link');
        moreButtons.forEach((moreButton) => {
          let parentOffer = moreButton.parentElement;
          while (!parentOffer.classList.contains('tbf-so-offer') &&
            !parentOffer.classList.contains('tbf-special-offers')) {
            parentOffer = parentOffer.parentElement;
          }
          moreButton.click();
          expect(parentOffer.classList.contains('tbf-so-open')).toBeTruthy();
          moreButton.click();
          expect(parentOffer.classList.contains('tbf-so-open')).not.toBeTruthy();
        });
      });
  });
});
