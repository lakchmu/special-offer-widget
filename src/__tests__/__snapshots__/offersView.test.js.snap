// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing class methods Testing default render function 1`] = `
<div
  class="special-offers tbf-so-special-offers"
>
  
  
  <div
    class="tbf-so-offer"
  >
    
    
    <div
      class="tbf-so-offer__content"
    >
      
      
      <div
        class="tbf-so-offer__content-image"
      >
        
        
        <img
          class="tbf-so-offer__image "
          src="https://s3.amazonaws.com/buuqit-dev/special_offers/images/000/000/009/public/Room2.jpg?1515606281"
        />
        
      
      </div>
      
      
      <div
        class="tbf-so-offer__content-text"
      >
        
        
        <div
          class="tbf-so-offer__title"
        >
          Last Minute Booking
        </div>
        
        
        <div
          class="tbf-so-offer__wrapper-description"
        >
          
          
          <div
            class="tbf-so-offer__short-description"
          >
            Last minute rate, Non-Refundable
          </div>
          
          
          <div
            class="tbf-so-offer__description"
          >
            Last minute booking discount, full pre payment required on time of booking. Please note this rate is non-refundable.
          </div>
          
        
        </div>
        
        
        <div
          class="tbf-so-offer__footer"
        >
          
          
          <span
            class="tbf-so-offer__dates"
          >
            
            
            <i
              class="tbf-so-offer__icon-calendar"
            />
            Jun 9 - Aug 31
          
          </span>
          
          
          <span
            class="tbf-so-offer__control-links"
          >
            
            
            <button
              class="tbf-so-offer__more-link"
            >
              More
            </button>
            
            
            <a
              class="tbf-so-offer__booking-link"
              href="https://app.thebookingfactory.com/Spilman-Hotel/book/__special_offer_9"
            >
              Book now
            </a>
            
          
          </span>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  


  
  <div
    class="tbf-so-offer"
  >
    
    
    <div
      class="tbf-so-offer__content"
    >
      
      
      <div
        class="tbf-so-offer__content-image"
      >
        
        
        <img
          class="tbf-so-offer__image "
          src="https://s3.amazonaws.com/buuqit-dev/special_offers/images/000/000/006/public/Room7.jpg?1512864349"
        />
        
      
      </div>
      
      
      <div
        class="tbf-so-offer__content-text"
      >
        
        
        <div
          class="tbf-so-offer__title"
        >
          Stay 3 Nights Save 15%
        </div>
        
        
        <div
          class="tbf-so-offer__wrapper-description"
        >
          
          
          <div
            class="tbf-so-offer__short-description"
          >
            Stay 3 Nights Save 15%
          </div>
          
          
          <div
            class="tbf-so-offer__description"
          >
            Stay for 3 nights or more and save 15%. Save a little by staying with us a bit longer
          </div>
          
        
        </div>
        
        
        <div
          class="tbf-so-offer__footer"
        >
          
          
          <span
            class="tbf-so-offer__dates"
          >
            
            
            <i
              class="tbf-so-offer__icon-calendar"
            />
            Jun 4 - Aug 31
          
          </span>
          
          
          <span
            class="tbf-so-offer__control-links"
          >
            
            
            <button
              class="tbf-so-offer__more-link"
            >
              More
            </button>
            
            
            <a
              class="tbf-so-offer__booking-link"
              href="https://app.thebookingfactory.com/Spilman-Hotel/book/__special_offer_6"
            >
              Book now
            </a>
            
          
          </span>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  

</div>
`;

exports[`Testing custom render function Testing rendering 1`] = `
<div
  class="special-offers tbf-so-special-offers"
>
  <div
    class="tbf-so-offer"
  >
    

    <div
      class="tbf-so-offer__title"
    >
      Last Minute Booking
    </div>
    

  </div>
  

  <div
    class="tbf-so-offer"
  >
    

    <div
      class="tbf-so-offer__title"
    >
      Stay 3 Nights Save 15%
    </div>
    

  </div>
</div>
`;
